server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/RangeHistogramAggregator.java

```diff
@@ -0,0 +1,169 @@
+    private final DocValueFormat formatter;
+    private final double interval, offset;
+    private final BucketOrder order;
+    private final boolean keyed;
+    private final long minDocCount;
+    private final double minBound, maxBound;
+
+    private final LongHash bucketOrds;
+
+    RangeHistogramAggregator(String name, AggregatorFactories factories, double interval, double offset,
+                               BucketOrder order, boolean keyed, long minDocCount, double minBound, double maxBound,
+                               @Nullable ValuesSource.Bytes.FieldData.RangeFieldData valuesSource, DocValueFormat formatter,
+                               SearchContext context, Aggregator parent,
+                               List<PipelineAggregator> pipelineAggregators, Map<String, Object> metaData) throws IOException {
+
+        super(name, factories, context, parent, pipelineAggregators, metaData);
+        if (interval <= 0) {
+            throw new IllegalArgumentException("interval must be positive, got: " + interval);
+        }
+        this.interval = interval;
+        this.offset = offset;
+        this.order = InternalOrder.validate(order, this);
+        this.keyed = keyed;
+        this.minDocCount = minDocCount;
+        this.minBound = minBound;
+        this.maxBound = maxBound;
+        this.valuesSource = valuesSource;
+        this.formatter = formatter;
+
+        bucketOrds = new LongHash(1, context.bigArrays());
+    }
+
+    @Override
+    protected LeafBucketCollector getLeafCollector(LeafReaderContext ctx, LeafBucketCollector sub) throws IOException {
+        if (valuesSource == null) {
+            return LeafBucketCollector.NO_OP_COLLECTOR;
+        }
+        final SortedBinaryDocValues values = valuesSource.bytesValues(ctx);
+        final RangeType rangeType = valuesSource.rangeType();
+        return new LeafBucketCollectorBase(sub, values) {
+            @Override
+            public void collect(int doc, long bucket) throws IOException {
+                assert bucket == 0;
+                if (values.advanceExact(doc)) {
+                    // Is it possible for valuesCount to be > 1 here? Multiple ranges are encoded into the same BytesRef in the binary doc
+                    // values, so it isn't clear what we'd be iterating over.
+                    final int valuesCount = values.docValueCount();
+
+                    double previousKey = Double.NEGATIVE_INFINITY;
+                    for (int i = 0; i < valuesCount; i++) {
+                        BytesRef encodedRanges = values.nextValue();
+                        // This list should be sorted by start-of-range, I think?
+                        List<RangeFieldMapper.Range> ranges = rangeType.decodeRanges(encodedRanges);
+                        for (RangeFieldMapper.Range range : ranges) {
+                            for (double value = rangeType.doubleValue(range.getFrom()); value <= rangeType.doubleValue(range.getTo());
```

Comment:
Perhaps pull the from/to out of the loop into temp variables, just to make the line a little shorter/easier to read?